src/react/Reconciler.js

```diff
@@ -0,0 +1,332 @@
+    if (
+      value instanceof StringValue ||
+      value instanceof NumberValue ||
+      value instanceof BooleanValue ||
+      value instanceof NullValue ||
+      value instanceof UndefinedValue
+    ) {
+      // terminal values
+      return value;
+    } else if (value instanceof AbstractValue) {
+      for (let i = 0; i < value.args.length; i++) {
+        value.args[i] = this._resolveDeeply(value.args[i], context, BranchStatus.NEW_BRANCH);
+      }
+      return value;
+    }
+    if (value instanceof ArrayValue) {
+      this._resolveFragment(value, context, branchStatus);
+      return value;
+    }
+    if (value instanceof ObjectValue && isReactElement(value)) {
+      let typeValue = Get(this.realm, value, "type");
+      let propsValue = Get(this.realm, value, "props");
+      let refValue = Get(this.realm, value, "ref");
+      if (typeValue instanceof StringValue) {
+        // terminal host component. Start evaluating its children.
+        if (propsValue instanceof ObjectValue) {
+          let childrenProperty = propsValue.properties.get("children");
+          if (childrenProperty) {
+            let childrenPropertyDescriptor = childrenProperty.descriptor;
+            // if the descriptor is undefined, the property is likely deleted, if it exists
+            // proceed to resolve the children
+            if (childrenPropertyDescriptor !== undefined) {
+              let childrenPropertyValue = childrenPropertyDescriptor.value;
+              invariant(childrenPropertyValue instanceof Value, `Bad "children" prop passed in JSXElement`);
+              let resolvedChildren = this._resolveDeeply(childrenPropertyValue, context, branchStatus);
+              childrenPropertyDescriptor.value = resolvedChildren;
+            }
+          }
+        }
+        return value;
+      }
+      // we do not support "ref" on <Component /> ReactElements
+      if (!(refValue instanceof NullValue)) {
+        this._assignBailOutMessage(value, `Bail-out: refs are not supported on <Components />`);
+        return value;
+      }
+      if (!(propsValue instanceof ObjectValue || propsValue instanceof AbstractValue)) {
+        return value;
+      }
+      if (!(typeValue instanceof ECMAScriptSourceFunctionValue)) {
+        return value;
+      }
+      try {
+        let { result, commitDidMountPhase } = this._renderAsDeepAsPossible(
+          typeValue,
+          propsValue,
+          context,
+          this._updateBranchStatus(branchStatus) // reduce branch status a value if possible
+        );
+        if (result === null || result instanceof UndefinedValue) {
+          return branchStatus === BranchStatus.NEW_BRANCH ? this._applyBranchedLogic(value) : value;
+        }
+        this.statistics.inlinedComponents++;
+        if (commitDidMountPhase !== null) {
+          commitDidMountPhase();
+        }
+        if (branchStatus === BranchStatus.NEW_BRANCH && result instanceof ObjectValue && isReactElement(result)) {
```

Comment:
If the result of this is a fragment or abstract value, then you don't apply the branch logic? Seems like that could lead to bugs. Be careful about dropping the "else" conditions for things like this.

Reply:
It should get applied, just not at this stage. Only ReactElements get covered at this stage and they are always `ObjectValue`s?